### PR TITLE
add bash tool use formatter

### LINKED ISSUE #920: format_bash

### PR DESCRIPTION
**Pull request by LasmarKhalifa**
closes https://github.com/Shopify/roast/issues/920

Improves the formatting of the bash tool use message.

Current output:

New output:

### CODE REVIEW

**Comment by LasmarKhalifa** [comment]
* **#945** <a href="https://app.graphite.com/github/pr/Shopify/roast/945?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#944** <a href="https://app.graphite.com/github/pr/Shopify/roast/944?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#943** <a href="https://app.graphite.com/github/pr/Shopify/roast/943?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#942** <a href="https://app.graphite.com/github/pr/Shopify/roast/942?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#941** <a href="https://app.graphite.com/github/pr/Shopify/roast/941?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#940** <a href="https://app.graphite.com/github/pr/Shopify/roast/940?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#939** <a href="https://app.graphite.com/github/pr/Shopify/roast/939?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#938** <a href="https://app.graphite.com/github/pr/Shopify/roast/938?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#937** <a href="https://app.graphite.com/github/pr/Shopify/roast/937?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#936** <a href="https://app.graphite.com/github/pr/Shopify/roast/936?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#935** <a href="https://app.graphite.com/github/pr/Shopify/roast/935?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#934** <a href="https://app.graphite.com/github/pr/Shopify/roast/934?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#933** <a href="https://app.graphite.com/github/pr/Shopify/roast/933?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/Shopify/roast/933?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: lib/roast/cogs/agent/providers/claude/tool_use.rb [hunk 1]
```diff
@@ -27,11 +27,37 @@ def format
               format_unknown
             end
 
+            TRUNCATE_LIMIT = 45
```
Comment (juniper-shopify) [inline lib/roast/cogs/agent/providers/claude/tool_use.rb:31]: This constant will be referenced across the entire formatter stack — a brief inline comment explaining the rationale behind the value would help future maintainers. Something like:

```ruby
# Truncate long formatted tool-use strings to keep terminal output to generally one line, accounting for logger prefixing.
TRUNCATE_LIMIT = 45
```

Small thing, but it saves the next person from wondering "why 45 and not 40 or 50 or 80?"

File: lib/roast/cogs/agent/providers/claude/tool_use.rb [hunk 2]
```diff
@@ -27,11 +27,37 @@ def format
...
+
             private
 
+            # Truncates a string to TRUNCATE_LIMIT characters, appending "..." when cut.
```
Comment (juniper-shopify) [inline lib/roast/cogs/agent/providers/claude/tool_use.rb]: The doc comment here is a good start, but since `truncate` is the shared workhorse for every formatter in this stack, it deserves the same level of detail you gave `format_bash`. Specifically:

```ruby
# Truncates a string to at most TRUNCATE_LIMIT characters, appending
# "..." when the string is cut. The returned string — including the
# ellipsis — is guaranteed to be ≤ TRUNCATE_LIMIT characters.
#
# nil is coerced to "" (so callers needn't nil-check).
# Strings at or below the limit are returned unchanged.
#
# Input:
#   str (String?) – the string to truncate
#
# Output: the (possibly shortened) string
#
# Examples:
#   truncate("ls -la")           # => "ls -la"
#   truncate("a" * 50)           # => "aaa...aaa..." (45 chars)
#   truncate(nil)                # => ""
```

The principle here: when a method is going to be used everywhere, investing in thorough docs pays dividends for anyone who needs to understand its boundary behavior without reading the implementation.

File: test/roast/cogs/agent/providers/claude/tool_use_test.rb [hunk 3]
```diff
@@ -15,13 +15,58 @@ class ToolUseTest < ActiveSupport::TestCase
...
+              assert_equal "#{"a" * (ToolUse::TRUNCATE_LIMIT - 3)}...", output
+            end
+
+            test "truncate handles nil without crashing" do
+              tool_use = ToolUse.new(name: :bash, input: {})
+
+              output = tool_use.send(:truncate, nil)
+
+              assert_equal "", output
+            end
```
Comment (juniper-shopify) [inline test/roast/cogs/agent/providers/claude/tool_use_test.rb]: A quick edge case test to add here alongside the existing truncate tests:

```ruby
test "truncate returns an empty string for an empty input" do
  tool_use = ToolUse.new(name: :bash, input: {})
  assert_equal "", tool_use.send(:truncate, "")
end
```

Even though `""` falls through the same path as `nil.to_s`, an explicit test documents that empty-string behavior is intentional rather than accidental.

File: lib/roast/cogs/agent/providers/claude/tool_use.rb [hunk 4]
```diff
@@ -27,11 +27,37 @@ def format
...
+            # Examples:
+            #   BASH ls -la (List directory contents)
+            #   BASH ls -la
+            #
             #: () -> String
             def format_bash
-              "BASH #{input.inspect}"
+              command = truncate(input[:command])
+              description = input[:description]
+              description ? "BASH #{command} (#{description})" : "BASH #{command}"
```
Comment (juniper-shopify) [inline lib/roast/cogs/agent/providers/claude/tool_use.rb]: why did you choose to truncate command and not description?

**Review by juniper-shopify — CHANGES_REQUESTED** [review]

File: test/roast/cogs/agent/providers/claude/tool_use_test.rb (hunk 3 above)
Comment (juniper-shopify) [reply test/roast/cogs/agent/providers/claude/tool_use_test.rb]: Also, when the truncated command ends up being "", the formatted result will be "BASH " with a trailing space. That's kind of messy.

File: lib/roast/cogs/agent/providers/claude/tool_use.rb (hunk 4 above)
Comment (LasmarKhalifa) [reply lib/roast/cogs/agent/providers/claude/tool_use.rb]: The command can be a very long string and doesn't add much detail, so that one should definitely be truncated.  
  
The description, on the other hand, is very useful and is quite short. The longest description we've had across 2000+ bash tool use logs is 102 characters and the average is ~40 characters. Since we don't expect it to be super long, and having an extra few words can be very helpful in understanding the goal behind a command, I took the decision that, for those rare cases where we have a slightly longer description, risking having an extra line wrapping around but getting a full understanding the goal was better than truncating halfway through and having the description lose much of its value.  
  
Not opposed to going either direction though, I see pros and cons to both truncating and not truncating the description.

File: test/roast/cogs/agent/providers/claude/tool_use_test.rb (hunk 3 above)
Comment (LasmarKhalifa) [reply test/roast/cogs/agent/providers/claude/tool_use_test.rb]: Added the case where we have a trailing whitespace for extra robustness!

**Review by juniper-shopify — APPROVED** [review]

**Comment by LasmarKhalifa** [comment]
### Merge activity

* **Jun 25, 6:03 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/Shopify/roast/933).
* **Jun 25, 6:03 PM UTC**: [Graphite](https://app.graphite.com/github/pr/Shopify/roast/933) rebased this pull request as part of a merge.
* **Jun 25, 6:04 PM UTC**: @LasmarKhalifa merged this pull request with [Graphite](https://app.graphite.com/github/pr/Shopify/roast/933).
